src/expr/src/relation/canonicalize.rs

```diff
@@ -18,7 +18,11 @@ use repr::{Datum, RelationType, ScalarType};
 /// This function:
 /// * ensures the same expression appears in only one equivalence class.
 /// * ensures the equivalence classes are sorted and dedupped.
-/// * simplifies expressions to involve the least number of non-leaves.
+/// * simplifies expressions to involve the least number of function calls
+///   involving one or more arguments. This is to prevent infinite loops/100%
+///   CPU situations where transforms can cause the equivalence class to evolve
+///   from `[A f(A) f(C, A)]` to `[A f(A) g(C, A) g(C, f(A))]` and then to
+///   `[A f(A) g(C, A) g(C, f(A)) g(C, f(f(A)))]` and so on.
```

Comment:
I'm sorry, I don't know what this means. The example doesn't make it as clear to me as I suspect it makes it clear to someone who has seen this problem before.

Reply:
Ok, more reading here, can I recommend the following text instead:
> This ensures that we only replace expressions by "even simpler" expressions. This ensures that repeated substitutions reduce the complexity of expressions and a fixed point is certain to be reached. Without this rule, we might repeatedly replace a simple expression with a complex expression containing that (or another replaceable) simple expression, and repeat indefinitely.